egressgw: cache matching endpoints in policy config

jibi opened the pull request:
Currently whenever we need to determine if a policy matches an endpoint, we call the `(*EndpointSelector).Matches()` method passing the target endpoint's labels.

Turns out this method is quite expensive so instead of recomputing each time if a policy matches a given endpoint, cache this information once for each `(*Manager)reconcile()` invocation, and then in `(*PolicyConfig)selectsEndpoint()` use the cached information.

jibi committed:
egressgw: cache matching endpoints in policy config

Currently whenever we need to determine if a policy matches an endpoint,
we call the (*EndpointSelector).Matches() method passing the target
endpoint's labels.

Turns out this method is quite expensive so instead of recomputing each
time if a policy matches a given endpoint, cache this information once
for each (*Manager)reconcile() invocation, and then in
(*PolicyConfig)selectsEndpoint() use the cached information.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

pchaigno reviewed (approved):
Second commit is mixing a fair big of refactoring/preparatory work with the actual changes (the four lines in `reconcile`), but otherwise LGTM.

jibi committed:
egressgw: pass event that triggered reconciliation to reconcile()

This will give the reconcile() method more context on the event that
triggered the reconciliation and will be used in a subsequent commit to
optimize part of that logic.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
egressgw: selectively update policies' matchedEndpointIDs

Instead of updating each policy's matchedEndpointIDs for each
reconciliation run, only update them:

* when a new policy is added (only update the single policy being added)
* when an endpoint is updated or deleted (in this case update all
  policies)

Signed-off-by: Gilberto Bertin <jibi@cilium.io>
